### PR TITLE
feat: add type branding to graphql in TS outputs

### PR DESCRIPTION
**Pull request by svidgen**
#### Description of changes

In order to [brand](https://typescript.tv/best-practices/improve-your-type-safety-with-branded-types/) the generated graphql for TypeScript, we need to know the underlying operation type (Query | Mutation | Subscription) and specific operation name (CreateBlog, ListBlogs, etc.) at the time the graphql is formatted for the language output. This PR:

1. Weaves op name and type through the `statements` map.
2. Derives relative types import path from config.
3. Conditionally adds types if the language is `typescript`.

#### Description of how you validated changes

1. Updated snapshot for new TS output (tests pass)
2. Ran through scenarios (To double-check functionality, these are combinations to replicate):

| library | codegen | expected outcome | |
| -- | -- | -- | -- |
| old | old | Base cases / no change | ✅  |
| old | new | Base cases / no change | ✅   |
| new | new | Base cases + API calls show specific arg and return types, preventing bad invocation | ✅   |
| new | old | Base cases + allows custom typing of both `arguments` and `result` | ✅   |

**Base cases**
1. Untyped graphql has generic typing &mdash; API category cannot infer or enforce specifics
2. Overrides [per the docs](https://docs.amplify.aws/lib/graphqlapi/query-data/q/platform/js/) using `GraphQLQuery` and `GraphQLSubscription` wrappers provides return types

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-codegen/blob/main/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] Breaking changes to existing customers are released behind a feature flag or major version update
- [ ] Changes are tested using sample applications for all relevant platforms (iOS/android/flutter/Javascript) that use the feature added/modified

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-codegen/src/utils/GraphQLStatementsFormatter.js [hunk 1]
```diff
@@ -36,27 +56,47 @@ class GraphQLStatementsFormatter {
...
-        formattedStatements.push(
-          `export const ${key} = /* GraphQL */ \`${value}\``
-        );
+      for (const [key, { graphql, operationName, operationType }] of statements) {
+        const typeTag = this.buildTypeTag(operationName, operationType);
+        formattedStatements.push(`export const ${key} = /* GraphQL */ \`${graphql}\`${typeTag}`);
       }
     }
-    const formattedOutput = [lintOverridesBuffer, headerBuffer, LINE_DELIMITOR, ...formattedStatements].join(LINE_DELIMITOR);
+    const formattedOutput = [lintOverridesBuffer, headerBuffer, LINE_DELIMITOR, this.typeDefs, LINE_DELIMITOR, ...formattedStatements].join(
```
[the remark this answers is not in the document]
Comment (alharris-at) [reply packages/amplify-codegen/src/utils/GraphQLStatementsFormatter.js:75]: And probably don't want to append the delimiter+TypeDef is this is language javascript.

**Review by alharris-at — APPROVED** [review]

File: packages/amplify-codegen/src/utils/GraphQLStatementsFormatter.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/amplify-codegen/src/utils/GraphQLStatementsFormatter.js:74]: `typeDefs` is implemented as a getter. As-is, this just results in an extra empty line if this isn't TypeScript, but I can clean that up.

Is the getter a frowned upon pattern here?

**Comment by svidgen** [comment]
Marking as draft again until PM's can review next Thursday. Happy to incorporate feedback between now and then though.

File: packages/amplify-codegen/src/utils/GraphQLStatementsFormatter.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (alharris-at) [reply packages/amplify-codegen/src/utils/GraphQLStatementsFormatter.js]: No I don't think the getter is necessarily frowned upon, it just wasn't obvious to me that would shortcircuit into a newline.

**Review by alharris-at — APPROVED** [review]

**Review by phani-srikar — APPROVED** [review]

**Pull request merged**